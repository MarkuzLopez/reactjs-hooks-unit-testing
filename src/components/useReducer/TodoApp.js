import React, { useEffect, useReducer } from 'react';
import { todoReducer } from './todoReducer';
import './styles.css';
import { useForm } from '../useEffect/useForm';
import TodoList from './TodoList';
import { TodoForm } from './TodoForm';

const init = () => { 

    return JSON.parse(localStorage.getItem('todos')) || [];
}


const TodoApp = () => {

    // const [ todos, dispatch ] = useReducer(todoReducer, initialState);
    
    // const [ formValues, hadleInputChange ] = useForm({
        //     description: ''
        // });
        
        // const [{ description }, hadleInputChange, resetForm]  = useForm({
            //     description: ''
            // });
            
            
                // const onSubmitTodo = async (e) => {
                //     e.preventDefault();
            
                //     if(!description.trim().length) { 
                //         return;
                //     }
            
                //     const newTodo = {
                //         id: new Date().getTime(),
                //         desc: description,
                //         done: false
                //     };
            
                //     const action = { 
                //         type: 'add',
                //         payload: newTodo
                //     }
            
                //     await dispatch(action)
                //     await resetForm();
                // }
                
    const [todos, dispatch] = useReducer(todoReducer, [], init);

    useEffect(() => {
        localStorage.setItem('todos', JSON.stringify(todos));
    }, [todos]); 
    

    const toggleTodo = (todoId) => { 
        dispatch({
            type: 'toggle',
            payload: todoId
        })
    }

    const hadleAddTodo = (newTodo) => { 
        dispatch({ 
            type: 'add',
            payload: newTodo
        })
    }

    const deleteTodo = (todoId) => { 
        //action
        const action = {
            type: 'delete',
            payload: todoId
        }

        //dispatch
        dispatch(action);
    }
    
    return (
        <div>
            <h1>TodoApp ( {todos.length} )</h1>
            <hr />
            <div className="row">
                <div className="col-8">
                    <TodoList
                        todosArray={todos}
                        toggleTodo={toggleTodo}
                        deleteTodo={deleteTodo}
                    />
                </div>
                <div className="col-4">
                    <TodoForm 
                     hadleAddTodo={hadleAddTodo}
                    />
                </div>
            </div>
        </div>

    );
}
 
export default TodoApp;