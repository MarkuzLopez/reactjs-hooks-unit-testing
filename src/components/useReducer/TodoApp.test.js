import React from 'react'
import { shallow, mount } from "enzyme"
import TodoApp from './TodoApp'
import { demoTodos } from '../../tests/fixtures/demosTodos';
import { act } from '@testing-library/react';

describe('Prueba unitaria TodoApp', () => {
    
    const wrapper = shallow(
        <TodoApp />
    );

    //obtienedo datos del storage
    Storage.prototype.setItem = jest.fn(() => {});

    test('deberia de realizar el snapshot ', () => {
       expect(wrapper).toMatchSnapshot(); 
    });
    
    test('debe de agregar un TODO', () => {
        
        const wrapper = mount(<TodoApp />)//engloba a todo el componente

        act(() => { 
            wrapper.find('TodoForm').prop('hadleAddTodo')(demoTodos[0]);
            wrapper.find('TodoForm').prop('hadleAddTodo')(demoTodos[1]);
        });
        // console.log(wrapper.find('h1').html());
        // expect(wrapper.find('h1').text().trim() ).toBe('TodoApp ( 2 )');
        expect(localStorage.setItem ).toHaveBeenCalledTimes(2);
    });

    test('deberia de eliminar un TODO', () => {
        
        wrapper.find('TodoForm').prop('hadleAddTodo')(demoTodos[0]);
        wrapper.find('TodoList').prop('deleteTodo')(demoTodos[0]);

        expect(wrapper.find('h1').text().trim() ).toBe('TodoApp ( 1 )');
    });
    
    
})
