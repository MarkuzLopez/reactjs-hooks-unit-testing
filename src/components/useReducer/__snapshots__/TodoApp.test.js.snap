// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba unitaria TodoApp deberia de realizar el snapshot  1`] = `
<div>
  <h1>
    TodoApp ( 
    0
     )
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-8"
    >
      <TodoList
        deleteTodo={[Function]}
        todosArray={Array []}
        toggleTodo={[Function]}
      />
    </div>
    <div
      className="col-4"
    >
      <TodoForm
        hadleAddTodo={[Function]}
      />
    </div>
  </div>
</div>
`;
